packages/cloud-functions/src/exchangeRate/FirebasePriceUpdater.ts

```diff
@@ -16,43 +14,123 @@ const MAX_CONCURRENCY = 30
 
 const RETRIES_LIMIT = 5
 
+interface ImageUrls {
+  [address: string]: string
+}
+
 export default class FirebasePriceUpdater {
   manager: ExchangeRateManager
 
   constructor(manager: ExchangeRateManager) {
     this.manager = manager
   }
 
-  async refreshAllPrices(): Promise<PriceByAddress> {
+  async updateAllPrices(): Promise<PriceByAddress> {
     const prices = await this.manager.calculatecUSDPrices()
     await this.updatePrices(prices)
     return prices
   }
 
   private async updatePrices(prices: PriceByAddress) {
+    const fetchTime = Date.now()
+
     const tokensInfoRaw = await fetchFromFirebase(FIREBASE_NODE_KEY)
 
-    const fetchTime = Date.now()
+    const imagesUrls = await this.getImagesUrlInfo()
+
+    await this.updatePresentTokens(prices, fetchTime, imagesUrls, tokensInfoRaw)
```

Comment:
If you want to keep the optimization we're discussing, I do think the code is clearer with one update function.  Pseudo-code, for example:

```ts
await function updateSingleTokenMetadata(...) {
  const record = {}
  if (doesntExistYet) {
    // call contract
   record.tokenContract = ...
   record.symbol = ...
   ...
  }
  record.whateverElse = ...
  await updateFirebase(`${FIREBASE_NODE_KEY}/${tokenAddress}`, record)
}

await asyncPool(MAX_CONCURRENCY, Object.values(tokensInfoRaw), ([key, token]) => {
  updateSingleTokenMetadata(...)
})

Reply:
Ok... I'll change it